magicgui/_tests/test_ qt.py

```diff
@@ -90,3 +101,75 @@ def test_set_categorical(qtbot):
     assert [w.itemText(i) for i in range(w.count())] == ["a", "b"]
     _qt.set_categorical_choices(w, (("a", 1), ("c", 3)))
     assert [w.itemText(i) for i in range(w.count())] == ["a", "c"]
+
+
+def test_magicfiledialog(qtbot):
+    """Test the MagicFileDialog class."""
+    filewidget = _qt.MagicFileDialog()
+
+    # check default values
+    assert filewidget.get_path() == Path('.')
+    assert filewidget.mode == _qt.FileDialogMode.OPTIONAL_FILE
+
+    # set the mode
+    filewidget.mode = _qt.FileDialogMode.EXISTING_FILES  # Enum input
+    assert filewidget.mode == _qt.FileDialogMode.EXISTING_FILES
+    filewidget.mode = "oPtioNal_FiLe"  # improper capitalization
+    assert filewidget.mode == _qt.FileDialogMode.OPTIONAL_FILE
+    filewidget.mode = "EXISTING_DIRECTORY"  # string input
+    assert filewidget.mode == _qt.FileDialogMode.EXISTING_DIRECTORY
+    with pytest.raises(ValueError):
+        filewidget.mode = 123  # invalid mode
+
+    # set the path
+    filewidget.set_path('my/example/path/')
+    assert filewidget.get_path() == Path('my/example/path/')
+
+    filewidget.mode = _qt.FileDialogMode.EXISTING_FILES
+    filewidget.set_path(['path/one.txt', 'path/two.txt'])
+    assert filewidget.get_path() == (Path('path/one.txt'), Path('path/two.txt'))
+    filewidget.set_path(['path/3.txt, path/4.txt'])
+    assert filewidget.get_path() == (Path('path/3.txt'), Path('path/4.txt'))
+
+    with pytest.raises(TypeError):
+        filewidget.set_path(123)  # invalid type, only str/Path accepted
+
+
+@pytest.mark.skipif("sys.platform == 'darwin'")  # dialog box hangs on Mac
+@pytest.mark.parametrize("mode", [
+    ('r'),  # existing file
+    ('EXISTING_DIRECTORY'),
+])
+def test_magicfiledialog_opens_chooser(qtbot, mode):
+    """Test the choose button opens a popup file dialog."""
+    filewidget = _qt.MagicFileDialog()
+    filewidget.set_path(('.',))  # set_path with tuple for better code coverage
+    filewidget.mode = mode
+
+    def handle_dialog():
+        popup_filedialog = next(
+            child
+            for child in filewidget.children()
+            if isinstance(child, QtW.QFileDialog)
+        )
+        assert isinstance(popup_filedialog, QtW.QFileDialog)
+        popup_filedialog.reject()
+
+    QtCore.QTimer().singleShot(400, handle_dialog)
+    filewidget._on_choose_clicked()
+
+
+@pytest.mark.skipif("sys.platform == 'windows'")  # Windows only test
+def test_windows_magicfiledialog(qtbot):
+    """Test we get a MagicFileDialog from a WindowsPath type."""
+    w = _qt.type2widget(WindowsPath)
+    assert w is not None
+    assert _qt.MagicFileDialog in w.mro()  # since type(w) = sip.wrappertype
+
+
+@pytest.mark.skipif("sys.platform != 'windows'")  # Linux and Mac test
+def test_linux_mac_magifiledialog(qtbot):
+    """Test we get a MagicFileDialog from a PosixPath type."""
+    w = _qt.type2widget(PosixPath)
+    assert w is not None
+    assert _qt.MagicFileDialog in w.mro()  # since type(w) = sip.wrappertype
```

Comment:
since we're not actually instantiating any of these classes, we can run this test on all pathtypes on all platforms.  also, `type2widget` is actually kinda poorly named :) it should really be `type2widgetclass` or something.  So we can actually just test for equality here.

```suggestion
@pytest.mark.parametrize("pathtype", [PosixPath, WindowsPath, Path])
def test_linux_mac_magifiledialog(pathtype):
    """Test we get a MagicFileDialog from a various Path types."""
    assert _qt.type2widget(pathtype) == _qt.MagicFileDialog
```

Reply:
This is beautiful and concise, I like your change a lot!